base_uom_converter/models/uom_converter.py

```diff
@@ -0,0 +1,139 @@
+from odoo import _, fields, models
+from odoo.exceptions import ValidationError
+
+
+class UomConverter(models.Model):
+    _name = "uom.converter"
+    _description = "UoM converter"
+
+    _sql_constraints = [("unique_name", "unique (name)", "This name already exists")]
+
+    name = fields.Char(string="Scale", required=True, index=True)
+    from_uom_id = fields.Many2one(
+        "uom.uom",
+        "Source Unit of Measure",
+        required=True,
+        help="Source unit of measure, while converting a quantity it "
+        "must be in this UoM to get the right scale. Or in a unit in the "
+        "same category that would be convert before getting the right line.",
+    )
+    to_uom_id = fields.Many2one(
+        "uom.uom",
+        "Destinaiton Unit of Measure",
+        required=True,
+        help="Destination unit of measure, while converting a quantity result "
+        "will be in this UoM. Or in a unit in the same category that would be "
+        "convert after getting the new quantity.",
+    )
+    line_ids = fields.One2many(
+        "uom.converter.line",
+        "uom_converter_id",
+        "UoM converter lines",
+    )
+
+    def convert(self, quantity, uom_qty=None, result_uom=None):
+        """Return quantity in to_uom_id or result_uom if set and in the same
+        uom category.
+
+        Given quantity is used to select the right scale line. if uom_qty is set
+        and different from the `self.from_uom_id` quantity it is converted
+        before scale line selection as long they share the same uom_category.
+        """
+        self.ensure_one()
+        if not uom_qty:
+            uom_qty = self.from_uom_id
+        else:
+            if uom_qty.category_id.id != self.from_uom_id.category_id.id:
+                raise ValidationError(
+                    _(
+                        "You can't convert {uom} (expected {category} uom "
+                        "category) to {to_uom} using this converter {converter}."
+                    ).format(
```

Comment:
please use fstrings.

Reply:
hum... I thought that the fstring were interpreted before the translation of the string, which would not be appropriate, I will test

Reply:
as expected:
 if I use that code
```diff
                 raise ValidationError(
                     _(
-                        "You can't convert {uom} to {to_uom} (expect "
-                        "{expected_category} unit category) using this "
-                        "converter {converter}."
-                    ).format(
-                        uom=uom_qty.name,
-                        to_uom=result_uom.name,
-                        expected_category=self.to_uom_id.category_id.name,
-                        converter=self.name,
+                        f"You can't convert {uom_qty.name} to {result_uom.name} (expect "
+                        f"{self.to_uom_id.category_id.name} unit category) using this "
+                        f"converter {self.name}."
                     )
                 )
```
I'm getting this error will translating the module

```
2021-04-09 10:07:04,450 12761 ERROR enrj_com_uom_base_converter odoo.tools.translate: Failed to extract terms from /home/pverkest/clients/phoenix/enrj-odoo/src/product-attribute/base_uom_converter/models/uom_converter.py 
Traceback (most recent call last):
  File "/home/pverkest/clients/phoenix/enrj-odoo/ocb/odoo/tools/translate.py", line 1118, in _babel_extract_terms
    for extracted in extract.extract(extract_method, src_file, keywords=extract_keywords, options=options):
  File "/home/pverkest/.pyenv/versions/3.7.10/envs/enrj-odoo/lib/python3.7/site-packages/babel/messages/extract.py", line 322, in extract
    for lineno, funcname, messages, comments in results:
  File "/home/pverkest/.pyenv/versions/3.7.10/envs/enrj-odoo/lib/python3.7/site-packages/babel/messages/extract.py", line 484, in extract_python
    value = eval(code, {'__builtins__': {}}, {})
  File "<string>", line 2, in <module>
NameError: name 'uom_qty' is not defined
```

Reply:
Strange, I did the following test: 
```py
$ odoo shell
from odoo import _
_(f"Hello { self.name } ") 
> 'Hello OdooBot '
```

Reply:
My traceback wasn't raising at runtime but while generating po/pot files.

Also, in your example `"Hello OdooBot "` will be interpreted before calling translation function `_()`  we don't want to translate each case for each users only the template string "Hello {something}" which will translate for any users. That's why I use `.format(` after the result of the translated string (`_()`).

Note that [odoo offer some facitilies](https://github.com/OCA/OCB/blob/14.0/odoo/tools/translate.py#L451) using the old deprecated `%-formatting` that we can call like this `_("Hello %", self.name)` but I avoided to depends on that assumption if odoo change that behavior in future versions.

Reply:
I was not aware of these issues with f-strings.

.format() have some security issues and should be avoided. 

% is an acceptable solution then

Reply:
I didn't know about security issue using `.format()` could you point me some references or tell me more about which attack to mitigate? the templated string is not set by end users I can't see why there are more risk than %-string ? :thinking:

Reply:
https://lucumr.pocoo.org/2016/12/29/careful-with-str-format/

Reply:
Thanks, oh yes of course users can access to translation string so I supposed we would get the same kind of issue with f-string by the way.

[string.Template](https://docs.python.org/3/library/string.html?highlight=template#string.Template) could be an alternative. I'll change to `%-string` !

Reply:
I suppose f-string are safe according to:
https://security.stackexchange.com/questions/238338/are-there-any-security-concerns-to-using-python-f-strings-with-user-input

Reply:
In fact I can't see an easy way to set a variable with an f-string template so that kind of security issue  (where end users are able to set the template string) is not possible !

To do something similar as funcionnality (let enduser set the template string) like in the last stackexchange would require an insecure eval likes:
```python
mavar = "{obj.name}"

class A:
  name="hello"
 
obj = A()

eval(f"f'{mavar}'")
'hello'

# or

mavar = "f'{obj.name}'"

eval(mavar)
'hello'
```
I was curious to make sure the odoo `safe_eval` mitigate f-string usage, it does:

```python
>>> from odoo.tools.safe_eval import safe_eval
>>> safe_eval('test', locals_dict={'test': "hello"})
'hello'
>>> safe_eval('f"{test}"', locals_dict={'test': "hello"})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/tools/safe_eval.py", line 328, in safe_eval
    c = test_expr(expr, _SAFE_OPCODES, mode=mode)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/tools/safe_eval.py", line 189, in test_expr
    assert_valid_codeobj(allowed_codes, code_obj, expr)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/tools/safe_eval.py", line 166, in assert_valid_codeobj
    raise ValueError("forbidden opcode(s) in %r: %s" % (expr, ', '.join(opname[x] for x in (code_codes - allowed_codes))))
ValueError: forbidden opcode(s) in 'f"{test}"': FORMAT_VALUE
```

Reply:
@hparfr thanks for this review, I've push 7ebb445 to use %-string instead format-string.